packages/dcos-integration-test/extra/test_mesos.py

```diff
@@ -457,3 +457,116 @@ def principal():
                 'unreserve_resources': resources}
             r = dcos_api_session.post('/mesos/api/v1', json=request)
             assert r.status_code == 202, r.text
+
+
+def reactivate_agent(dcos_api_session, agent_id):
+    request = {
+        'type': 'REACTIVATE_AGENT',
+        'reactivate_agent': {
+            'agent_id': {
+                'value': agent_id
+            }
+        }
+    }
+
+    r = dcos_api_session.post('/mesos/api/v1', json=request)
+    assert r.status_code == 200, r.text
+
+
+def test_node_drain_and_reactivate(dcos_api_session):
+    app, test_uuid = test_helpers.marathon_test_app()
+    app_id = 'integration-test-{}'.format(test_uuid)
+
+    with dcos_api_session.marathon.deploy_and_cleanup(app):
+        # Fetch the mesos master state once the task is running
+        master_ip = dcos_api_session.masters[0]
+        r = dcos_api_session.get('/state', host=master_ip, port=5050)
+        assert r.status_code == 200
+        state = r.json()
+
+        # Find the agent ID from master state
+        agent_id = None
+        for framework in state['frameworks']:
+            for task in framework['tasks']:
+                if app_id in task['id']:
+                    agent_id = task['slave_id']
+        assert agent_id is not None, 'Agent ID not found for instance of app_id {}'.format(app_id)
+
+        # Drain the agent
+        request = {
+            'type': 'DRAIN_AGENT',
+            'drain_agent': {
+                'agent_id': {'value': agent_id}
+            }
+        }
+
+        r = dcos_api_session.post('/mesos/api/v1', json=request)
+        assert r.status_code == 200, r.text
+
+        try:
+            # Verify that the agent is drained. The agent will not transition to
+            # DRAINED immediately, so we retry
+            @retrying.retry(wait_fixed=1000, stop_max_delay=10000)
+            def get_state_and_verify_drained():
+                r = dcos_api_session.get('/state', host=master_ip, port=5050)
+                assert r.status_code == 200
+                state = r.json()
+
+                agent_found = False
+                for agent in state['slaves']:
+                    if agent['id'] == agent_id:
+                        agent_found = True
+                        assert 'drain_info' in agent, 'Agent {} does not include drain info'.format(agent_id)
+                        assert agent['drain_info']['state'] == 'DRAINED'
+
+                assert agent_found
+
+            get_state_and_verify_drained()
+        except Exception as ex:
```

Comment:
Let's add a comment that this is part of the test teardown and needs to be invoked unconditionally. This would be clearer if the `try` block would contain all our mutations to the cluster (starting agent draining, confirm agent is draining, reactivating agent).

A more scalable alternative would be to use a proper test fixutre with a tear down function.

Here and below.